dodiscover/ci/clf_test.py

```diff
@@ -0,0 +1,359 @@
+from typing import Callable, Optional, Set, Tuple, Union
+
+import numpy as np
+import pandas as pd
+import scipy.special
+import sklearn
+import sklearn.metrics
+from numpy.typing import NDArray
+from sklearn.neighbors import NearestNeighbors
+from sklearn.utils import check_random_state
+
+from dodiscover.typing import Column
+
+from .base import BaseConditionalIndependenceTest
+
+
+class ClassifierCITest(BaseConditionalIndependenceTest):
```

Comment:
Seems very closely related to https://arxiv.org/abs/1804.02747, which is also implemented (in a significantly improved form) here: https://github.com/py-why/dowhy/blob/main/dowhy/gcm/independence_test/regression.py#L14

A general problem with these ML prediction based approaches is, that they don't find all kind of dependencies, only when they impact the expectation. For instance, a dependency with respect to the variance:

```
import numpy as np

from dowhy.gcm import kernel_based, regression_based

X = np.random.normal(0, 1, 1000)
Y = []

for x in X:
    Y.append(np.random.normal(0, abs(x)))

Y = np.array(Y)
Z = np.random.normal(0, 1, 1000)

print("Correct result:", kernel_based(X, Y, Z))
print("Wrong result", regression_based(X, Y, Z))

clf = RandomForestClassifier()
ci_estimator = ClassifierCITest(clf)

df = pd.DataFrame({'x': X, 'y': Y, 'z': Z})

_, pvalue = ci_estimator.test(df, {"x"}, {"z"}, {"y"})
print("Wrong result", pvalue)
```

Here, ```X``` and ```Y``` are dependent (even after conditioning on ```Z```, which is irrelevant here), but only the kernel based test would find this dependence, while the ML based test would claim they are independent.

Nevertheless, it is still worth having it implemented :)

Reply:
This is a great point. I'll note it in the docs and provide an example showing the counter-example. I guess I can do that in this PR.

Reply:
> A general problem with these ML prediction based approaches is, that they don't find all kind of dependencies, only when they impact the expectation. For instance, a dependency with respect to the variance:

I don't suppose this is a huge problem for dependence induced by _causality_.  I'm trying to think of an example of an SCM where a causal parent would only affect the variance of the child.  Maybe a if X and Y were parents of Z, Z is a sensor reading, X is the true signal, and Y is a source of interference in the sensor reading?